### PR TITLE
Provide a way for extensions to discover used classpath entries

### PR DESCRIPTION
**Pull request by guw**
(Patch for #2529)

## Author checklist

- [x] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/NameEnvironmentAnswerWithClasspathOrigin.java
Comment (guw) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/NameEnvironmentAnswerWithClasspathOrigin.java:12]: @jukzi I don't like a separate class. I wonder if we can have another interface next to `NameEnvironmentAnswer` and just add this field there. All I need is the `Path` of the jar or directory (in terms of batch compiler). But I can image that an IDE would want to have more interesting objects here.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/FileSystem.java
Comment (guw) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/FileSystem.java:169]: We probably need a better type here and getter/setter?

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/NameEnvironmentAnswerWithClasspathOrigin.java
Comment (iloveeclipse) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/NameEnvironmentAnswerWithClasspathOrigin.java:12]: > I wonder if we can have another interface next to `NameEnvironmentAnswer` (like `IClasspathOrigin`) and just add this field directly in `NameEnvironmentAnswer`

Sounds good.

Comment (iloveeclipse) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/NameEnvironmentAnswerWithClasspathOrigin.java:12]: That would also remove unneeded change in jdt.core

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/FileSystem.java
Comment (iloveeclipse) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/batch/FileSystem.java:169]: I would not make it over complicated, `Consumer<NameEnvironmentAnswer>` is OK for what is needed.
What I would add is the only public method `Consumer<NameEnvironmentAnswer> setNameEnvironmentAnswerListener(Consumer<NameEnvironmentAnswer>) and make this field private, as I don't like public fields. Return value would be original listener, if any.

**Comment by iloveeclipse** [comment]
What would be also nice is a simple dedicated test that compiles a class and checks that the name listener "sees" the expected class path. Example tests that create compiler were org.eclipse.jdt.core.tests.compiler.regression.AnnotationTest.test265() or org.eclipse.jdt.core.tests.compiler.regression.EnclosingMethodAttributeTest.test004().

**Comment by stephan-herrmann** [comment]
@guw one wild idea: have you seen that `NameEnvironmentAnswer` already has a field `AccessRestriction accessRestriction`?

Could that field be used for your purpose? As of now you would need to set an access rule to make all paths "discouraged", but perhaps one more problemId would suffice so you can write access rules that cause collecting "access restrictions" which then, however, will not raise a problem, but only serve for data collection. WDYT?

**Comment by guw** [comment]
I will evaluate a simple change relying on `IDependent.getFileName()`.

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NameEnvironmentAnswerListenerTest.java
Comment (iloveeclipse) [inline org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NameEnvironmentAnswerListenerTest.java:39]: The test must be added to existing test suite.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
In general OK, few issues below.

Also commit message `Provide ability to track unused dependencies` doesn't match https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

I will push all required changes in a moment.

**Comment by iloveeclipse** [comment]
@guw : looks like you've created PR from the master branch and don't allow collaboration on master branch either.
Could you please:
1) pull from origin and rebase your branch on master
2) apply attached patch via `git am < PR_2530_patch.txt` to the PR

[PR_2530_patch.txt](https://github.com/user-attachments/files/15816138/PR_2530_patch.txt)

**Comment by guw** [comment]
@iloveeclipse Is the patch against latest master and not this PR? It does not apply cleanly.

**Comment by iloveeclipse** [comment]
> @iloveeclipse Is the patch against latest master and not this PR? It does not apply cleanly.

Sure. As I've wrote, you have to rebase on master first.

**Comment by iloveeclipse** [comment]
You could make this PR editable for contributors, so I can push directly to your branch.

**Comment by guw** [comment]
@iloveeclipse I updated the PR with your recommendations. Which setting is it to make PRs editable? I thought you had to fork my repo and then push a PR agains my repo's main branch.

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NameEnvironmentAnswerListenerTest.java
Comment (iloveeclipse) [inline org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NameEnvironmentAnswerListenerTest.java:170]: Could we please use my version?
```
		EclipseBatchCompiler compiler;
		File logFile = new File(outputDirectory, "compile.log");
		try(PrintWriter log = new PrintWriter(new FileOutputStream(logFile))) {
			compiler = new EclipseBatchCompiler(log);
			boolean compileOK;
			compileOK = compiler.compile(ecjArguments.toArray(new String[ecjArguments.size()]));
			if(!compileOK) {
				String logOutputString = Util.fileContent(logFile.getAbsolutePath());
				Assert.fail("Compile failed, output: '" + logOutputString + "'");
			}
		}
```
With the current one, if you break compilation, test reports only `java.lang.AssertionError: Compile failed but did not throw exception!` but in my version we would see immediately **why** :
```
java.lang.AssertionError: Compile failed, output: '----------
1. ERROR in /tmp/comptest/run.1718287287470/lib/src/s/X.java (at line 1)
	pack age s;
	^^^^^^^^
Syntax error on tokens, they can be merged to form package
----------
1 problem (1 error)```

**Comment by guw** [comment]
@iloveeclipse I was able to figure out how to apply the patch. Instead of "rebase" my branch on master I simply reset it. This allowed applying your patch.

**Comment by iloveeclipse** [comment]
Last patch turned leading tabs to spaces in all files except Main :-(

**Comment by iloveeclipse** [comment]
@guw : could you please fix the patch?

**Comment by guw** [comment]
@iloveeclipse I am a bit lost. I did apply the patch with `git am`. Not sure why it would screw up the spaces/tabs.

**Comment by iloveeclipse** [comment]
Just fix them manually, you can see it in the diff. I have no idea what broke whitespace, the patch contains tabs...
Next time would be really good to have a PR where others could push, I would have fixed/merge everything already yesterday.

**Comment by guw** [comment]
That's blocked by GitHub not me: https://github.com/orgs/community/discussions/5634

**Comment by iloveeclipse** [comment]
> That's blocked by GitHub not me: https://github.com/orgs/community/discussions/5634

Not sure if that's different case.
Is https://github.com/eclipseguru really an "organization"??? Or is it just the fact that "master" branch (on which this PR is created) is always read-only for anyone else except repo owner?

**Comment by iloveeclipse** [comment]
Gunnar, it is so sad to see the failing attempts to fix the code.

Now spaces are gone, but almost all new code has extra tab indentation level, plus `TestAll` got unneeded whitespace changes in unrelated code & an extra commit is added that isn't needed (should be all in one commit).

Should I create PR in my repo from my last patch? It would save us a day or two at least :-)

**Comment by guw** [comment]
Indeed. JDT builds all these things things for Eclipse like format on save but JDT doesn't use them. This is too much toil. No wonder there is a lack of contributors. 🤷  My last attempt was AnyEdit tools and use "Replace Spaces with Tabs" command.

FWIW, I only opened the PR upon request. I don't have a specific interest in seeing this merged. It is easer for me to just maintain a fork. Please feel free to do whatever necessary to get it merged.

**Comment by iloveeclipse** [comment]
> FWIW, I only opened the PR upon request. 

I really appreciate that, and I encourage you to avoid **patching** JDT in your code but if you see some code that needs some improvement to be useful, either create a ticket or push a PR immediately (but not from the master branch next time).

> I don't have a specific interest in seeing this merged. It is easer for me to just maintain a fork. Please feel free to do whatever necessary to get it merged.

@guw : I think the main issue with this PR that it was initially created from the "master" branch of your fork, so I wasn't able to help you with your attempts to fix it.

I've created https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2577 based on the patch I've added here before. Once it builds and will be green, I plan to merge it.

**Pull request closed without merging**
